incubator/hnc/internal/validators/hierarchy.go

```diff
@@ -210,9 +212,55 @@ func (v *Hierarchy) checkParent(ns, curParent, newParent *forest.Namespace) admi
 		return deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
 	}
 
+	// Prevent overwriting source objects in the descendants after the hierarchy change.
+	if yes, co := v.hasConflictObject(newParent, ns); yes {
+		coStr := strings.Join(co, "\n  * ")
+		msg := fmt.Sprintf("\nCannot update hierarchy because it would overwrite the following object(s):\n  * %s\nTo fix this, please rename or remove the conflicting objects first.", coStr)
+		return deny(metav1.StatusReasonConflict, msg)
+	}
+
 	return allow("")
 }
 
+// hasConflictObject returns true and a list of objects if there's a conflict.
+func (v *Hierarchy) hasConflictObject(pns, cns *forest.Namespace) (bool, []string) {
+	// Get all types with 'Propagate' mode.
+	ts := []forest.TypeSyncer{}
+	for _, t := range v.Forest.GetTypeSyncers() {
+		if t.GetMode() == api.Propagate {
+			ts = append(ts, t)
+		}
+	}
+
+	// Get all the original objects in the new ancestors (including the parent).
+	existsInAns := make(map[schema.GroupVersionKind]map[string]bool)
```

Comment:
Maybe `newSrcObjs`?

Reply:
Updated with `newAnsSrcObjs`?

Reply:
Are there ancestor objects that can't be sources? Just wondering why we need both descriptions. Also "Ans" is a bit of a hard short form to read, while both "src" and "obj" are fairly well-understood.

Reply:
Because we want to compare the source objects in all NEW ancestors for a parent change, so it's not just source objects in one namespace but new ancestors.